src/refract/JSONSchemaVisitor.cc

```diff
@@ -304,73 +293,75 @@ namespace refract
+            bool allEmpty = allItemsEmpty(val);
+
+            for (ArrayElement::ValueType::const_iterator it = val->begin();
+                 it != val->end();
+                 ++it) {
+
+                if (*it) {
+                    // if all items are just type items then we
+                    // want them in the schema, otherwise skip
+                    // empty ones
+                    if (allEmpty || !(*it)->empty()) {
+                        JSONSchemaVisitor v(true);
+
+                        v.visit(*(*it));
+                        av.push_back(v.getOwnership());
                     }
                 }
+            }
 
-                if (!av.empty()) {
-                    ArrayElement *a = new ArrayElement;
-                    a->set(av);
-                    addMember("items", a);
-                }
+            if (!av.empty()) {
+                ArrayElement *a = new ArrayElement;
+                a->set(av);
+                addMember("items", a);
             }
+        }
 
-            if (IsTypeAttribute(e, "required")) {
-                for (ArrayElement::ValueType::const_iterator it = val->begin();
-                     it != val->end();
-                     ++it) {
+        if (IsTypeAttribute(e, "required")) {
+            for (ArrayElement::ValueType::const_iterator it = val->begin();
+                 it != val->end();
+                 ++it) {
 
-                    if (*it && !(*it)->empty()) {
-                        renderer.visit(*(*it));
-                    }
+                if (*it && !(*it)->empty()) {
+                    renderer.visit(*(*it));
                 }
-
-                addMember("items", renderer.getOwnership());
             }
+
+            addMember("items", renderer.getOwnership());
         }
 
         const ArrayElement *def = GetDefault(e);
 
         if (def && !def->empty()) {
-            if (e.element() == "enum") {
-                IElement *d = def->value[0]->clone();
-                d->renderType(IElement::rCompact);
-                addMember("default", d);
-            }
-            else {
-                ArrayElement *d = static_cast<ArrayElement*>(def->clone());
-                d->renderType(IElement::rCompact);
-                for (ArrayElement::ValueType::iterator i = d->value.begin();
-                     i != d->value.end();
-                     ++i) {
-                    (*i)->renderType(IElement::rCompact);
-                }
-                addMember("default", d);
-            }
+            ArrayElement *d = static_cast<ArrayElement*>(def->clone());
+            d->renderType(IElement::rCompact);
+            SetRenderFlag(d->value, IElement::rCompact);
+            addMember("default", d);
+        }
+    }
+
+    void JSONSchemaVisitor::visit(const EnumElement& e) {
+        const EnumElement::ValueType* val = GetValue<EnumElement>(e);
+
+        if (!val || val->empty()) {
+            return;
+        }
+
+        enumElement(e, val);
```

Comment:
Could not these two methods be merged, it is now unnecessary to have it as it was just to clarify the `visit(ArrayElement)` method ? It would make it simpler and avoid unnecessary double `GetDefault(e)` call.